### PR TITLE
fix: Treat float -0.0 == 0.0 and -NaN == NaN in group-by, joins and unique

### PR DESCRIPTION
**Pull request by nameexhaustion**
Fixes https://github.com/pola-rs/polars/issues/10457

Before / After
```python
# Before
>>> df.group_by("x", maintain_order=True).agg("index")
shape: (4, 2)
┌──────┬───────────┐
│ x    ┆ index     │
│ ---  ┆ ---       │
│ f64  ┆ list[u32] │
╞══════╪═══════════╡
│ -0.0 ┆ [0]       │
│ 0.0  ┆ [1]       │
│ NaN  ┆ [2]       │
│ NaN  ┆ [3]       │
└──────┴───────────┘
# After
>>> df.group_by("x", maintain_order=True).agg("index")
shape: (2, 2)
┌──────┬───────────┐
│ x    ┆ index     │
│ ---  ┆ ---       │
│ f64  ┆ list[u32] │
╞══════╪═══════════╡
│ -0.0 ┆ [0, 1]    │
│ NaN  ┆ [2, 3]    │
└──────┴───────────┘
```

I took a bit of interest in trying this out, though I wasn't able to stop after I started 😅

### CODE REVIEW

File: crates/polars-utils/src/total_ord.rs [hunk 1]
```diff
@@ -402,3 +436,140 @@ impl<T: TotalOrd, U: TotalOrd> TotalOrd for (T, U) {
...
+}
+
+impl<'a> TotalEq for BytesHash<'a> {
+    fn tot_eq(&self, other: &Self) -> bool {
+        self == other
+    }
+}
+
+/// This elides creating a [`TotalOrdWrap`] for types that don't need it.
+pub trait ToTotalOrd {
```
[the remark this answers is not in the document]
Comment (orlp) [reply crates/polars-utils/src/total_ord.rs:456]: Could you change it to use `TotalOrdWrap`? That's precisely its purpose.

**Comment by ritchie46** [comment]
The reproducable hash test may be overwritten. The values don't matter as long as its constant within the same Polars version.

File: crates/polars-core/src/chunked_array/object/mod.rs [hunk 2]
```diff
@@ -36,7 +37,7 @@ pub trait PolarsObjectSafe: Any + Debug + Send + Sync + Display {
 
 /// Values need to implement this so that they can be stored into a Series and DataFrame
 pub trait PolarsObject:
-    Any + Debug + Clone + Send + Sync + Default + Display + Hash + PartialEq + Eq + TotalEq
+    Any + Debug + Clone + Send + Sync + Default + Display + Hash + TotalHash + PartialEq + Eq + TotalEq
```
Comment (ritchie46) [inline crates/polars-core/src/chunked_array/object/mod.rs:40]:
```suggestion
    Any + Debug + Clone + Send + Sync + Default + Display + TotalHash + PartialEq + TotalEq
```

File: crates/polars-ops/src/frame/join/hash_join/single_keys_semi_anti.rs [hunk 3]
```diff
@@ -35,7 +40,8 @@ fn semi_anti_impl<T, I>(
 ) -> impl ParallelIterator<Item = (IdxSize, bool)>
 where
     I: IntoIterator<Item = T> + Copy + Send + Sync,
-    T: Send + Hash + Eq + Sync + Copy + DirtyHash,
```
[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-ops/src/frame/join/hash_join/single_keys_semi_anti.rs:38]: Is there a lint that checks this?

File: crates/polars-utils/src/total_ord.rs [hunk 4]
```diff
@@ -149,6 +152,37 @@ impl<T: Clone> Clone for TotalOrdWrap<T> {
...
+        self.0.is_null()
+    }
+
+    fn unwrap_inner(self) -> Self::Inner {
+        self.0.unwrap_inner()
+    }
+}
+
+impl DirtyHash for f32 {
+    fn dirty_hash(&self) -> u64 {
```
Comment (ritchie46) [inline crates/polars-utils/src/total_ord.rs:169]: Can we tag this with inline?

File: crates/polars-utils/src/total_ord.rs [hunk 5]
```diff
@@ -149,6 +152,37 @@ impl<T: Clone> Clone for TotalOrdWrap<T> {
...
+        canonical_f32(*self).to_bits().dirty_hash()
+    }
+}
+
+impl DirtyHash for f64 {
+    fn dirty_hash(&self) -> u64 {
```
Comment (ritchie46) [inline crates/polars-utils/src/total_ord.rs:175]: Can we tag this with inline?

File: crates/polars-utils/src/total_ord.rs [hunk 6]
```diff
@@ -149,6 +152,37 @@ impl<T: Clone> Clone for TotalOrdWrap<T> {
...
+        canonical_f64(*self).to_bits().dirty_hash()
+    }
+}
+
+impl<T: DirtyHash> DirtyHash for TotalOrdWrap<T> {
+    fn dirty_hash(&self) -> u64 {
```
Comment (ritchie46) [inline crates/polars-utils/src/total_ord.rs:181]: Can we tag this with inline?

File: crates/polars-utils/src/total_ord.rs [hunk 7]
```diff
@@ -402,3 +436,18 @@ impl<T: TotalOrd, U: TotalOrd> TotalOrd for (T, U) {
             .then_with(|| self.1.tot_cmp(&other.1))
     }
 }
+
+impl<'a> TotalHash for BytesHash<'a> {
+    fn tot_hash<H>(&self, state: &mut H)
```
Comment (ritchie46) [inline crates/polars-utils/src/total_ord.rs:441]: Can we tag this with inline?

File: crates/polars-utils/src/total_ord.rs [hunk 8]
```diff
@@ -402,3 +436,18 @@ impl<T: TotalOrd, U: TotalOrd> TotalOrd for (T, U) {
...
+    where
+        H: Hasher,
+    {
+        self.hash(state)
+    }
+}
+
+impl<'a> TotalEq for BytesHash<'a> {
+    fn tot_eq(&self, other: &Self) -> bool {
```
Comment (ritchie46) [inline crates/polars-utils/src/total_ord.rs:450]: Can we tag this with inline?

**Review by ritchie46 — COMMENTED** [review]
Great one! We already had the architecture for this. But great that we now finally made it correct for the most important algorithms.

I've left a few minor comments.

File: crates/polars-utils/src/total_ord.rs [hunk 9]
```diff
@@ -278,6 +320,7 @@ impl_float_eq_ord!(f32);
 impl_float_eq_ord!(f64);
 
 impl TotalHash for f32 {
+    #[inline(always)]
```
[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-utils/src/total_ord.rs:323]: I wouldn't inline it as `always`. Just `inline` and let the optimizer decide.

**Comment by ritchie46** [comment]
I am going to change this to a bug fix. Nobody should have depended on this behavior. This was clearly a bug, just one that was hard to fix.

**Review by ritchie46 — APPROVED** [review]

**Comment by ritchie46** [comment]
Ah.. I think we should if we can. They are faster right?

**Pull request merged**